File: labs/L15/2303/PrefixTree.py
class PrefixTree:

    def __init__(self, key=None):
        self.key = key
        self.leaf = False
        self.hasChildren = False
        self.children = [None] * 10

    def add(self, branch):
        curr = self
        for c in branch:
            if curr.leaf:
                return False

            key = int(c)
            if curr.children[key] == None:
                curr.hasChildren = True
                curr.children[key] = PrefixTree(key)

            curr = curr.children[key]

        if curr.leaf or curr.hasChildren:
            return False

        curr.leaf = True

        return True


with open("input.txt") as f:
    testNum = int(f.readline())
    for i in range(testNum):
        phones = PrefixTree()
        phonesNum = int(f.readline())
        res = True
        for ph in range(phonesNum):
            phone = (f.readline()).strip()
            if not res: continue

            res = phones.add(phone)

        if res:
            print("YES")
        else:
            print("NO")

File: labs/L15/2303/PrefixTree.cpp
#ifndef _CRT_SECURE_NO_WARNINGS
#define _CRT_SECURE_NO_WARNINGS

#include <stdio.h>

#endif

#include<iostream>

using namespace std;

static int testNum;

class PrefixTree {

public:
	PrefixTree()
		: leaf(false)
	{
		init();
	}

	void init() {
		for (int i = 0; i < 10; i++) {
			children[i] = 0;
		}
	}

	bool hasChildren() {
		for (int i = 0; i < 10; i++) {
			if (children[i] != 0) return true;
		}
		return false;
	}

	bool add(char* branch) {
		PrefixTree* curr = this;
		char* c = branch;
		while (*c != 0) {
			if (curr->leaf)
				return false;

			int key = *c - '0';
			if (curr->children[key] == 0) {
				curr->children[key] = new PrefixTree();
			}

			curr = curr->children[key];
			c++;
		}

		if (curr->leaf || curr->hasChildren()) {
			return false;
		}

		curr->leaf = true;
		return true;
	}

private:
	bool leaf;
	PrefixTree* children[10];
};

int main() {
    freopen("input.txt", "r", stdin);

    scanf("%d", &testNum);
    for (int i = 0; i < testNum; ++i) {
		PrefixTree book;
		int phonesNum;
		bool res = true;
        scanf("%d", &phonesNum);
        for (int ph = 0; ph < phonesNum; ++ph) {

            char phone[11];
            scanf("%s", &phone);
			if (!res) continue;
			res = book.add(phone);
        }

		if (res)
			printf("YES\n");
		else
			printf("NO\n");

    }
    return 0;
}